### PR TITLE
Make MLflow project runner supporting submit spark job to databricks runtime >= 13

### PR DESCRIPTION
**Pull request by WeichenXu123**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/WeichenXu123/mlflow/pull/12139?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12139/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12139
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Make MLflow project job containing spark jobs supporting databricks runtime >= 13

Context:

`mlflow run -b databricks ...` CLI command creates a databricks job cluster, and then MLflow submit the current directory project to Databricks cluster to execute.

In databricks runtime 10.x, the MLflow project code can run spark job in the created Databricks cluster, but after Databricks runtiem >= 13.x, this no longer works, because:

`mlflow run -b databricks ...` creates process in child process of databricks driver, and run user code in child process (**because in databricks runtime user code is launched by `mlflow run` shell command too, and further more, in shell command process, it creates another child process for using the created virtual python environment**), but in Databricks runtime >= 13.x, child process of spark driver is not allowed to create spark job in the databricks spark cluster.

To address this, we have to add new logic in MLflow project runner to make it supporting this following way:
when executing user code in Databricks, it doesn't create subprocess, instead, use https://docs.databricks.com/en/workflows/jobs/jobs-2.0-api.html#sparkpythontask . And for python dependencies, install them as databricks cluster libraries. Note the job have to use the builtin python version of the databricks runtime.

`MLProject` file configuration example:
```
name: tutorial

databricks_spark_job:
    python_file: "train.py"  # the file to be executed as the pyspark driver process
    parameters: []  # parameters, only allow a list of string
    python_libraries: # dependencies to be installed as databricks cluster libraries
      - mlflow==2.4.1
      - scikit-learn
      - pandas
      - dynaconf
      - numpy
```

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [X] Manual tests
Testing project: https://github.com/WeichenXu123/mlflow-job-test

Project 1: `mlflow-job-with-spark` it is configured as a **MLflow databricks spark job** .(See  `databricks_spark_job` section in MLproject file)
Project 2: `mlflow-job-without-spark` it is configured as legacy MLflow job.

To start above jobs, just enter the project directory, then run command like:
```
export MLFLOW_TRACKING_URI=databricks

mlflow run .  -b databricks --backend-config cluster-spec.json --experiment-id 648780469063303
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [x] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by WeichenXu123** [comment]
Our field engineer tested with this, it addressed customer's use-case.

File: mlflow/projects/_project_spec.py [hunk 1]
```diff
@@ -42,6 +42,39 @@ def load_project(directory):
...
+
+        if "entry_points" in yaml_obj:
+            raise MlflowException("databricks spark job does not support setting 'entry_points'.")
+
+        if env_type.DOCKER in yaml_obj:
+            raise MlflowException(
+                "databricks spark job does not support setting docker environment."
+            )
+
+        if env_type.PYTHON in yaml_obj:
```
Comment (BenWilson2) [inline mlflow/projects/_project_spec.py:58]: Is this a breaking change?

File: mlflow/projects/_project_spec.py [hunk 2]
```diff
@@ -42,6 +42,39 @@ def load_project(directory):
 
     project_name = yaml_obj.get("name")
 
+    databricks_spark_job_yaml = yaml_obj.get("databricks_spark_job")
+    if databricks_spark_job_yaml is not None:
+        if "python_file" not in databricks_spark_job_yaml:
+            raise MlflowException("'python_file' field is required for databricks spark job.")
```
Comment (BenWilson2) [inline mlflow/projects/_project_spec.py:48]: Can we captitalize "Databricks" and "Spark" in these error messages?

**Comment by BenWilson2** [comment]
Please file an update to https://docs.databricks.com/en/mlflow/projects.html that describes the changes here for MLR 13.x+ with regards to setting environment configurations. Since this is a breaking change that needs to happen to make this API usable in recent runtimes, let's ensure that a new section at the top of the docs, with appropriate warnings about compatibility and which configurations are not permitted any longer, is very clear to users.

**Review by BenWilson2 — APPROVED** [review]
LGTM after nit addressing. Let's be sure to get the docs PR filed this week to meet the next release train.

File: mlflow/projects/_project_spec.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/projects/_project_spec.py:58]: No.

all these checkings are only triggered when `if databricks_spark_job_yaml is not None` condition is satisfied.
i.e. only if user adds  `databricks_spark_job` section (new added feature) in the MLproject file, it triggers these checkings.

**Comment by WeichenXu123** [comment]
> LGTM after nit addressing. Let's be sure to get the docs PR filed this week to meet the next release train.

I will file a follow-up PR for this !

**Pull request merged**